cmd/errorformat/main.go

```diff
@@ -101,32 +105,66 @@ func run(r io.Reader, w io.Writer, efms []string, entryFmt, name string, list bo
 		efms = f.Errorformat
 	}
 
-	out := newTrackingWriter(w)
+	var writer Writer
 
-	fm := template.FuncMap{
-		"join": strings.Join,
-	}
-	tmpl, err := template.New("main").Funcs(fm).Parse(entryFmt)
-	if err != nil {
-		return err
+	switch writerFmt {
+	case "template", "":
+		fm := template.FuncMap{
+			"join": strings.Join,
+		}
+		tmpl, err := template.New("main").Funcs(fm).Parse(entryFmt)
+		if err != nil {
+			return err
+		}
+		writer = &TemplateWriter{Template: tmpl, Writer: newTrackingWriter(w)}
+	case "checkstyle":
+		writer = &CheckStyleWriter{w: w}
+	default:
+		return fmt.Errorf("unknown writer: -w=%v", writerFmt)
 	}
+	defer func() {
+		if err := writer.Flash(); err != nil {
+			log.Println(err)
+		}
+	}()
 
 	efm, err := errorformat.NewErrorformat(efms)
 	if err != nil {
 		return err
 	}
 	s := efm.NewScanner(r)
 	for s.Scan() {
-		if err := tmpl.Execute(out, s.Entry()); err != nil {
+		if err := writer.Write(s.Entry()); err != nil {
 			return err
 		}
-		if out.NeedNL() {
-			out.WriteNL()
-		}
 	}
 	return nil
 }
 
+type Writer interface {
+	Write(*errorformat.Entry) error
+	Flash() error
+}
+
+type TemplateWriter struct {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type TemplateWriter should have comment or be unexported